### PR TITLE
Add a pod stage for failing an initContainer 

### PR DESCRIPTION
**Pull request by yuanchen8911**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:

This PR adds a pod stage that injects an error to initContainer specified by a label and annotations, including the initContainer's name and optional failure message, reason, exit code and delay parameters.

```
labels:
   pod-init-container-running-failed.stage.kwok.x-k8s.io: 'true'
annotations:
   pod-init-container-running-failed.stage.kwok.x-k8s.io/container-name: <failed initContainer name>
   // optional
   pod-init-container-running-failed.stage.kwok.x-k8s.io/reason: <failure reason, the default is "initContainerError">
   pod-init-container-running-failed.stage.kwok.x-k8s.io/message: <failure message, the default is "initContainer reported errors" > 
   pod-init-container-running-failed.stage.kwok.x-k8s.io/exit-code: <exit code, the default is 1>
   pod-init-container-running-failed.stage.kwok.x-k8s.io/delay: <delay in milliseconds> 
   pod-init-container-running-failed.stage.kwok.x-k8s.io/jitter-delay: <jitter delay in milliseconds> 
```

#### Which issue(s) this PR fixes:

Fixes #
None
#### Special notes for your reviewer:

Questions:

1. How to set the default values for durationFrom and jitterDurationFrom
2. Do we need to set the statues of the other initContainers after the failed initContainer?

#### Does this PR introduce a user-facing change?

```release-note
Support the failure of an initContainer in a pod.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: kustomize/stage/pod/chaos/pod-init-container-failed.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      {{ range $index, $item := .spec.containers }}
+      - image: {{ $item.image | Quote }}
+        name: {{ $item.name | Quote }}
+        ready: false
+        restartCount: 0
+        started: false
+        state:
+          waiting:
+            reason: PodInitializing
+      {{ end }}
```
Comment (wzshiming) [inline kustomize/stage/pod/chaos/pod-init-container-failed.yaml:84]: It didn't work and I made some changes

- There's something wrong with JQ and Go Template syntax.
- This is already running the init container, so conditions and podIP cannot be omitted.

```diff
diff --git a/kustomize/stage/pod/chaos/pod-init-container-failed.yaml b/kustomize/stage/pod/chaos/pod-init-container-failed.yaml
index aaa0a53f..aaef7126 100644
--- a/kustomize/stage/pod/chaos/pod-init-container-failed.yaml
+++ b/kustomize/stage/pod/chaos/pod-init-container-failed.yaml
@@ -8,7 +8,7 @@ spec:
     kind: Pod
   selector:
     matchExpressions:
-    - key: '.metadata.labels.pod-init-container-failed.stage.kwok.x-k8s.io'
+    - key: '.metadata.labels["pod-init-container-running-failed.stage.kwok.x-k8s.io"]'
       operator: 'In'
       values:
       - 'true'
@@ -20,34 +20,43 @@ spec:
       - 'Pending'
   delay:
     durationFrom:
-      expressionFrom: '.metadata.annotations.pod-init-container-failed.stage.kwok.x-k8s.io/delay'
+      expressionFrom: '.metadata.annotations["pod-init-container-running-failed.stage.kwok.x-k8s.io/delay"]'
     jitterDurationFrom:
-      expressionFrom: '.metadata.annotations.pod-init-container-failed.stage.kwok.x-k8s.io/jitter-delay'
+      expressionFrom: '.metadata.annotations["pod-init-container-running-failed.stage.kwok.x-k8s.io/jitter-delay"]'
+  weight: 10000
   next:
-    delete: false
     statusTemplate: |
       {{ $now := Now }}
       {{ $defaultReason := "initContainerError" }}
       {{ $defaultMessage := "initContainer reported errors" }}
       {{ $defaultExitCode := 1 }}
-      {{ $initContainerName := .metadata.annotations.pod-init-container-failed.stage.kwok.x-k8s.io/container-name }}
-      {{ $failureReason := or .metadata.annotations.pod-init-container-failed.stage.kwok.x-k8s.io/reason $defaultReason}}
-      {{ $failureMessage := or .metadata.annotations.pod-init-container-failed.stage.kwok.x-k8s.io/message, $defaultMessage}}
-      {{ $failureExitCode := or .metadata.annotations.pod-init-container-failed.stage.kwok.x-k8s.io/exit-code $defaultExitCode}}
+      {{ $initContainerName := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/container-name" ) "" }}
+      {{ $failureReason := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/reason" ) $defaultReason }}
+      {{ $failureMessage := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/message" ) $defaultMessage }}
+      {{ $failureExitCode := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/exit-code" ) $defaultExitCode }}
+      
       conditions:
       - lastProbeTime: null
         lastTransitionTime: {{ $now | Quote }}
         status: "False"
         reason: ""
         type: Initialized
+      - lastTransitionTime: {{ $now | Quote }}
+        status: "False"
+        reason: ""
+        type: Ready
+      - lastTransitionTime: {{ $now | Quote }}
+        status: "False"
+        reason: ""
+        type: ContainersReady
+      
       initContainerStatuses:
       {{ range $index, $item := .spec.initContainers }}
-      {{ $origin := index $root.status.initContainerStatuses $index }}
-      {{ if eq $item.name $initContainerName }}
+      {{ if or ( not $initContainerName ) ( eq $item.name $initContainerName ) }}
       - image: {{ $item.image | Quote }}
         name: {{ $item.name | Quote }}
         ready: false
-        restartCount: {{ $origin.restartCount }} 
+        restartCount: 0
         started: false
         state:
           terminated:
@@ -56,13 +65,11 @@ spec:
             reason: {{ $failureReason }}
             message: {{ $failureMessage }}
             startedAt: {{ $now | Quote }}
-      phase: Failed
-      {{ break }}
       {{ else }}
       - image: {{ $item.image | Quote }}
         name: {{ $item.name | Quote }}
-        restartCount: {{ $origin.restartCount }} 
-        started: false
+        ready: true
+        restartCount: 0
         state:
           terminated:
             exitCode: 0
@@ -71,6 +78,7 @@ spec:
             startedAt: {{ $now | Quote }}
       {{ end }}
       {{ end }}
+      
       containerStatuses:
       {{ range $index, $item := .spec.containers }}
       - image: {{ $item.image | Quote }}
@@ -82,3 +90,8 @@ spec:
           waiting:
             reason: PodInitializing
       {{ end }}
+      
+      hostIP: {{ NodeIPWith .spec.nodeName | Quote }}
+      podIP: {{ PodIPWith .spec.nodeName ( or .spec.hostNetwork false ) ( or .metadata.uid "" ) ( or .metadata.name "" ) ( or .metadata.namespace "" ) | Quote }}
+      phase: Failed
+      startTime: {{ $now | Quote }}
```

```suggestion
apiVersion: kwok.x-k8s.io/v1alpha1
kind: Stage
metadata:
  name: pod-init-container-running-failed
spec:
  resourceRef:
    apiGroup: v1
    kind: Pod
  selector:
    matchExpressions:
    - key: '.metadata.labels["pod-init-container-running-failed.stage.kwok.x-k8s.io"]'
      operator: 'In'
      values:
      - 'true'
    - key: '.metadata.deletionTimestamp'
      operator: 'DoesNotExist'
    - key: '.status.phase'
      operator: 'In'
      values:
      - 'Pending'
  delay:
    durationFrom:
      expressionFrom: '.metadata.annotations["pod-init-container-running-failed.stage.kwok.x-k8s.io/delay"]'
    jitterDurationFrom:
      expressionFrom: '.metadata.annotations["pod-init-container-running-failed.stage.kwok.x-k8s.io/jitter-delay"]'
  weight: 10000
  next:
    statusTemplate: |
      {{ $now := Now }}
      {{ $defaultReason := "initContainerError" }}
      {{ $defaultMessage := "initContainer reported errors" }}
      {{ $defaultExitCode := 1 }}
      {{ $initContainerName := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/container-name" ) "" }}
      {{ $failureReason := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/reason" ) $defaultReason }}
      {{ $failureMessage := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/message" ) $defaultMessage }}
      {{ $failureExitCode := or ( index .metadata.annotations "pod-init-container-running-failed.stage.kwok.x-k8s.io/exit-code" ) $defaultExitCode }}
      
      conditions:
      - lastProbeTime: null
        lastTransitionTime: {{ $now | Quote }}
        status: "False"
        reason: ""
        type: Initialized
      - lastTransitionTime: {{ $now | Quote }}
        status: "False"
        reason: ""
        type: Ready
      - lastTransitionTime: {{ $now | Quote }}
        status: "False"
        reason: ""
        type: ContainersReady
      
      initContainerStatuses:
      {{ range $index, $item := .spec.initContainers }}
      {{ if or ( not $initContainerName ) ( eq $item.name $initContainerName ) }}
      - image: {{ $item.image | Quote }}
        name: {{ $item.name | Quote }}
        ready: false
        restartCount: 0
        started: false
        state:
          terminated:
            exitCode: {{ $failureExitCode }}
            finishedAt: {{ $now | Quote }}
            reason: {{ $failureReason }}
            message: {{ $failureMessage }}
            startedAt: {{ $now | Quote }}
      {{ else }}
      - image: {{ $item.image | Quote }}
        name: {{ $item.name | Quote }}
        ready: true
        restartCount: 0
        state:
          terminated:
            exitCode: 0
            finishedAt: {{ $now | Quote }}
            reason: Completed
            startedAt: {{ $now | Quote }}
      {{ end }}
      {{ end }}
      
      containerStatuses:
      {{ range $index, $item := .spec.containers }}
      - image: {{ $item.image | Quote }}
        name: {{ $item.name | Quote }}
        ready: false
        restartCount: 0
        started: false
        state:
          waiting:
            reason: PodInitializing
      {{ end }}
      
      hostIP: {{ NodeIPWith .spec.nodeName | Quote }}
      podIP: {{ PodIPWith .spec.nodeName ( or .spec.hostNetwork false ) ( or .metadata.uid "" ) ( or .metadata.name "" ) ( or .metadata.namespace "" ) | Quote }}
      phase: Failed
      startTime: {{ $now | Quote }}
```

File: kustomize/stage/pod/chaos/pod-init-container-failed.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
+apiVersion: kwok.x-k8s.io/v1alpha1
+kind: Stage
+metadata:
+  name: pod-init-container-running-failed
```
Comment (wzshiming) [inline kustomize/stage/pod/chaos/pod-init-container-failed.yaml:4]: Rename the file name to match this one.

IMO, one should distinguish between creating and running failed.

**Comment by yuanchen8911** [comment]
Thanks for the fixing, @wzshiming! All the changes were committed. I still have two questions.

1. Would it be possible to use some default values if the duration/jitterDuration annotations are not specified?
2.  The statues of those initContainers after the failed initContainers should NOT be `terminated` with `Completed`, right?

**Comment by wzshiming** [comment]
> Would it be possible to use some default values if the duration/jitterDuration annotations are not specified?

Sure, set both `durationMilliseconds` and `durationFrom.expressionFrom`, the `durationMilliseconds` will like a default for it.

> The statues of those initContainers after the failed initContainers should NOT be terminated with Completed, right?

Yes, the completed reason is only for the 0 exit code 

https://github.com/containerd/containerd/blob/406e9e84b4fb0a6bf2912f56ff1dfe78a56ed5ee/internal/cri/server/container_status.go#L89-L95

File: kustomize/stage/pod/chaos/pod-init-container-failed.yaml (hunk 2 above)
Comment (yuanchen8911) [reply kustomize/stage/pod/chaos/pod-init-container-failed.yaml]: Renamed the file to `pod-init-container-running-failed.yaml`.

**Comment by yuanchen8911** [comment]
> > Would it be possible to use some default values if the duration/jitterDuration annotations are not specified?
> 
> Sure, set both `durationMilliseconds` and `durationFrom.expressionFrom`, the `durationMilliseconds` will like a default for it.
Set both `durationMilliseconds` and `jitterDurationMillisecons` to 1000. (no jitter).
> 
> > The statues of those initContainers after the failed initContainers should NOT be terminated with Completed, right?
> 
> Yes, the `Completed` reason is only for the 0 exit code
> 
> https://github.com/containerd/containerd/blob/406e9e84b4fb0a6bf2912f56ff1dfe78a56ed5ee/internal/cri/server/container_status.go#L89-L95
My point was we should distinguish an initContainers defined before the failed initContainers and an one after the failed initContainer. The former's state is`terminated` while the latter's is `waiting`.

**Review by wzshiming — COMMENTED** [review]
/lgtm

Thank you, I think it's a good start for chaos testing

**Comment by yuanchen8911** [comment]
> /lgtm
> 
> Thank you, I think it's a good start for chaos testing

Removed the WIP label. Can you approve it again? Thanks!

**Comment by yuanchen8911** [comment]
> New changes are detected. LGTM label has been removed.

Fixed the linkyaml issues (trailing spaces). PTAL, @wzshiming. Thanks!

**Pull request merged**